### ISSUE TITLE
Model binding: Automatically bind dates in JS format?

### ISSUE THREAD

**Issue by strohhut** [aspnet/Mvc#7391]
It seems that per default dates in JavaScript format don't bind to DateTime. 

I have an ajax post:
``` JavaScript
$.post({
    url: "@Url.Action("MyAction", "Home")",
    cache: false,
    dataType: "json",
    data: {
        date1: new Date().toString(), date2: new Date(), date3: new Date().toISOString()
    }
}
```

And a view model:
``` CSharp
public class Class1
{
    public DateTime Date1 { get; set; }
    public DateTime Date2 { get; set; }
    public DateTime Date3 { get; set; }
}
```

And an action:
``` CSharp
[HttpPost]        
public IActionResult MyAction([FromForm]Class1 data)
{
    return new JsonResult(null);
}
```

This gets submitted:
```
date1=Mon+Feb+19+2018+20:10:28+GMT+0100
date2=Mon+Feb+19+2018+20:10:28+GMT+0100
date3=2018-02-19T19:10:28.165Z
```

Only `Date3` gets bound. Maybe it would make sense that `Date1` and `Date2` get bound out of the box too?

In my specific use case I wanted to get the data object that jQuery posts from knockout's `ko.mapping.toJS`. I have to do a workaround (e.g. using `data: JSON.parse(ko.mapping.toJSON(...))`) in order to make the dates bind because they are all Date objects and not ISO date strings.

**Comment by mkArtakMSFT** [comment]
@dougbu , @rynowak any thoughts regarding the above ask? Is there any restriction preventing us from trying to parse dates in that format? (assuming JavaScript `Date.toString()` produces the same format across browsers).

**Comment by dougbu** [comment]
The `Date.ToString()` method [apparently](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toString#Description) "always returns a string representation of the date in American English" and seems to use a format .NET Framework does not support.

We *could* implement a custom `Date` string parser and associated `IModelBinder` / `IModelBinderProvider` pair, similar to `EnumTypeModelBinder` and `EnumTypeModelBinderProvider`. I wouldn't recommend this because (a) `toISOString()` is a simple, supported workaround and (b) none of the existing datatype-specific model binders would be a tenth as complicated. In particular, the existing datatype-specific model binders use additional BCL support for those datatypes beyond the `TypeConverter` defaults.

If `toISOString()` is objectionable for some reason, users can add their own `IModelBinder` / `IModelBinderProvider` pairs to their applications.

**Comment by strohhut** [comment]
It's not always so easy to call toISOString() like in my case where knockout gives the view model object that will be posted via jQuery. And this object contains Date type properties. That's why I showed my toJSON/JSON.parse roundtrip workaround.

No data binding for JS format is not a bug or a critical missing feature but I think it would be nice to have. I spent some time wondering why the dates don't get bound and then to figure out the workaround. Supporting JS format might save developers some time and I think it's a common use case that dates are objects and not ISO strings and you don't know all the date properties that will get posted.

It would be possible to write JS functionality that transforms all date properties to ISO strings on the posted object but I think that would add unnecessary complexity.
